drivers/clock_control/clock_control_pwm.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2023 Andriy Gelman <andriy.gelman@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT pwm_clock
+
+#include <stdint.h>
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/pwm.h>
+#include <zephyr/dt-bindings/pwm/pwm.h>
+#include <zephyr/kernel.h>
+
+#define LOG_LEVEL CONFIG_CLOCK_CONTROL_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(clock_control_pwm);
+
+BUILD_ASSERT(CONFIG_CLOCK_CONTROL_PWM_INIT_PRIORITY > CONFIG_PWM_INIT_PRIORITY,
+	     "PWM must have a higher priority than PWM clock control");
+
+struct clock_control_pwm_config {
+	int num_pwms;
+	const struct pwm_dt_spec *pwm_dt;
+};
+
+struct clock_control_pwm_data {
+	uint32_t is_enabled_mask;
+	uint32_t *clock_frequency;
+};
+
+static int clock_control_pwm_on(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct clock_control_pwm_data *data = dev->data;
+	const struct clock_control_pwm_config *config = dev->config;
+	const struct pwm_dt_spec *spec;
+	int id = (int)sys;
+	int ret;
+
+	if (id >= config->num_pwms) {
+		return -EINVAL;
+	}
+
+	spec = &config->pwm_dt[id];
+	if (data->clock_frequency[id] == 0) {
+		ret = pwm_set_dt(spec, spec->period, spec->period / 2);
+	} else {
+		uint64_t cycles_per_sec;
+		uint32_t period_cycles;
+
+		ret = pwm_get_cycles_per_sec(spec->dev, spec->channel, &cycles_per_sec);
+		if (ret) {
+			return ret;
+		}
+
+		if (cycles_per_sec % data->clock_frequency[id] > 0) {
+			LOG_WRN("Target clock frequency cannot be expressed in PWM clock ticks");
+		}
+
+		period_cycles = cycles_per_sec / data->clock_frequency[id];
+		ret = pwm_set_cycles(spec->dev, spec->channel, period_cycles, period_cycles / 2,
+				     spec->flags);
+	}
+
+	if (ret) {
+		return ret;
+	}
+
+	data->is_enabled_mask |= BIT(id);
```

Comment:
Does this need locking?

Reply:
Yes, it looks that writing to it needs a lock.

Reply:
I simplified the driver to only support one PWM. It simplifies the macros and there's no need for locking now.
If the user needs more than one clock source, they can just create multiple instances of pwm-clock.